dwave/optimization/include/dwave-optimization/array.hpp

```diff
@@ -237,338 +237,473 @@ struct Update {
-
-    // Create a masked iterator with a fill value. Will return the value pointed at by *fill_ptr
-    // when *mask_ptr evaluates to true.
-    ArrayIterator(const value_type* data_ptr, const value_type* mask_ptr,
-                  const value_type* fill_ptr) noexcept
-            : ptr_(data_ptr),
-              mask_(std::make_unique<MaskInfo>(mask_ptr, fill_ptr)),
-              shape_(nullptr) {}
-
-    // shape and strides must outlive the iterator!
-    ArrayIterator(value_type const* ptr, ssize_t ndim, const ssize_t* shape, const ssize_t* strides)
-            : ptr_(ptr),
-              mask_(nullptr),
-              shape_((ndim >= 1) ? std::make_unique<ShapeInfo>(ndim, shape, strides) : nullptr) {}
-
-    // Both the copy and move operator using the copy-and-swap idiom
-    ArrayIterator& operator=(ArrayIterator other) noexcept {
-        using std::swap;  // ADL, if it matters
-        std::swap(ptr_, other.ptr_);
-        std::swap(mask_, other.mask_);
-        std::swap(shape_, other.shape_);
-        return *this;
-    }
-
-    const value_type& operator*() const {
-        if (mask_ && *(mask_->mask_ptr)) {
-            return *(mask_->fill_ptr);
+// Represents an Array
+//
+// This interface is designed to work with Python's buffer protocol
+// https://docs.python.org/3/c-api/buffer.html
+// We however, only implement a subset of the features. Specifically:
+//  * We only support doubles
+//
+// We also support a notion of state-dependent size. This allows nodes to
+// change their size based on the state of the decision variables.
+// Nodes are only permitted to grow and shrink along axis 0. This means that
+// growing or shrinking a node is equivalent to extending or shrinking the
+// buffer - there are no insertions needed.
+// Nodes will signal that they have a state-dependent size by returning negative
+// values for size() and in the first element of shape().
+class Array {
+ private:
+    // The implementation of an iterator over the Array.
+    //
+    // Arrays can be contiguous or strided, so an iterator for them must be
+    // too.
+    //
+    // For a contiguous array, the array iterator only needs to hold a pointer
+    // to the memory it is iterating over.
+    //
+    // For a strided array, the iterator also holds information about
+    // the strides and shape of the parent array, as well as its location within
+    // that array.
+    //
+    // For a masked array, the iterator also holds information about whether
+    // the current value should be masked, and what value to return if so.
+    //
+    // `IsConst` toggles whether the iterator is an output iterator or not.
+    template <bool IsConst>
+    class ArrayIteratorImpl_ {
+     public:
+        using difference_type = std::ptrdiff_t;
+        using value_type = std::conditional<IsConst, const double, double>::type;
+        using pointer = value_type*;
+        using reference = value_type&;
+
```

Comment:
Both `pointer` and `reference` are unused in this class as far as I can tell, but Im guessing they are intended to be used from outside the class for convenience?

Reply:
Yeah, this is a holdover from early iterator interface requirements. At one point C++ wanted iterators to have all of these types declared in their namespace. Though they also wanted [iterator_traits](https://en.cppreference.com/w/cpp/iterator/iterator_traits) which I haven't used.

In practice, most of the iterator implementations I've seen around still put them in, so I am inclined to leave them as harmless, if a bit redundant.